core/domain/state_domain.py

```diff
@@ -1595,6 +1604,12 @@ def validate(self):
         """
         self.feedback.validate()
 
+        if self.dest_if_really_stuck is not None:
```

Comment:
Hm, don't you need more validation than this (and corresponding backend tests for the various cases)? E.g.:

- The dest_if_really_stuck should actually be a valid state name.
- If the answer group is marked as "correct" then there shouldn't be a dest_if_really_stuck
- The dest_if_really_stuck shouldn't match the name of the state it is currently being used for
- When checking that every state is reachable from the start, you'd need to consider the dest_if_really_stuck links as well (this is in addition to what I imagine are existing checks to ensure that, with only the regular dest links, every state can reach the end).

I might have missed some, so please think through the different cases and ensure there are tests for these. Better to make this robust now so that you don't get people creating invalid explorations which we need to clean up later.

/cc @iamprayush

Reply:
Right, all the arguments make sense. Actually:

- A regex check would ensure that the state name entered is a valid one. Also, since the creator chooses the dest_if_really_stuck state from a dropdown or creates a new one, only valid states will be assigned to this field.
- In M1.2, a check is added that the dest_if_really_stuck UI become visible only if the answer group is not marked as correct.
- The same state name wouldn't appear in the dropdown. So, the case of a self-loop would be checked in the frontend itself.
- Actually, the check for this can only be implemented in M1.3 when new types of links would be created to connect the dest_if_really_stuck state.
Basically, these conditions will be applied in the frontend itself.
-

Reply:
Note that all the validation checks have to be complete in the backend though. The backend should be robust to any data coming through the handlers, because it's technically possible for data to come to the backend handler from somewhere other than the frontend (e.g. a rogue POST request).

Reply:
Added the validation checks to ensure that dest_if_really_stuck (in both outcome and default_outcome):
1. is a string if not null
2. should not be present when labelled_as_correct is true.
3. is a valid state
Also, added tests for the same. The fourth point is only possible in M1.3. PTAL @seanlip